### PR TITLE
phpDoc cleanup + automation

### PR DESCRIPTION
**Pull request by allejo**
At the moment, the phpDocs for individual functions, README, and the main `load.php` class are maintained manually and separately. What if we would make use of `phpdocumentor/reflection-docblock` to generate the documentation for us? Here's a sample of what was generated with the `phpDocGen.php` script I put together; this could easily be adapted to generate the README too.

```php
/**
 *
 *
 * @method static array append(array $param, mixed|null $param) <p>Append item to array</p><br /><pre><code class="language-php">__::append([1, 2, 3], 4);<br />  → [1, 2, 3, 4]</code></pre>
 * @method static array chunk(array $param, int $param, bool $param) <p>Creates an array of elements split into groups the length of size. If array can't be split evenly, the final chunk will be the remaining elements.</p><br /><p>__::chunk([1, 2, 3, 4, 5], 3);<br />// → [[1, 2, 3], [4, 5]]</p>
 * @method static array compact(array $param) <p>Creates  an  array  with  all  falsey  values removed. The values<br />false, null, 0, &quot;&quot;, undefined, and NaN are all falsey.</p><br /><ul><br /><li>__::compact([0, 1, false, 2, '', 3]);</li><br /><li>// → [1, 2, 3]</li><br /></ul>
 * @method static array drop(array $param, int $param) <p>Creates a slice of array with n elements dropped from the beginning.</p><br /><ul><br /><li>__::drop([0, 1, 3], 2);</li><br /><li>// → [3]</li><br /></ul>
 * @method static array flatten(void $param, bool $param) <p>Flattens a multidimensional array. If you pass shallow, the array will only be flattened a single level.</p><br /><p>__::flatten([1, 2, [3, [4]]], [flatten]);</p><br /><blockquote><br /><blockquote><br /><p>[1, 2, 3, 4]</p><br /></blockquote><br /></blockquote>
 * @method static array patch(array $param, array $param, string $param) <p>Patches array by xpath.</p><br /><ul><br /><li>__::patch(['addr' =&gt; ['country' =&gt; 'US', 'zip' =&gt; 12345]], ['/addr/country' =&gt; 'CA', '/addr/zip' =&gt; 54321]);</li><br /><li>// → ['addr' =&gt; ['country' =&gt; 'CA', 'zip' =&gt; 54321]]</li><br /></ul>
 * @method static array prepend(array $param, null $param) <p>prepend item or value to an array</p><br /><ul><br /><li>__::prepend([1, 2, 3], 4);</li><br /><li>// → [4, 1, 2, 3]</li><br /></ul>
 * @method static array randomize(array $param) <p>Shuffle an array ensuring no item remains in the same position.</p><br /><ul><br /><li>__::randomize([1, 2, 3]);</li><br /><li>// → [2, 3, 1]</li><br /></ul>
 * @method static array range(int|null $param, int|null $param, int $param) <p>generate range of values based on start , end and step</p><br /><ul><br /><li>__::range(1, 10, 2);</li><br /><li>// → [1, 3, 5, 7, 9]</li><br /></ul><br />
 * @method static array repeat(string $param, int $param) <p>generate array of repeated values</p><br /><ul><br /><li>__::repeat('foo', 3);</li><br /><li>// → ['foo', 'foo', 'foo']</li><br /></ul>
 * @method static array|object assign(array|object $param, array|object $param) <p>Combines and merge collections provided with each others.</p><br /><p>If the collections have common keys, then the last passed keys override the<br />previous. If numerical indexes are passed, then last passed indexes override<br />the previous.</p><br /><p>For a recursive merge, see <code>__::merge()</code>.</p><br /><pre><code>__::assign(<br />    ['color' =&gt; ['favorite' =&gt; 'red', 5], 3],<br />    [10, 'color' =&gt; ['favorite' =&gt; 'green', 'blue']]<br />);<br /><br />['color' =&gt; ['favorite' =&gt; 'green', 'blue'], 10]</code></pre>
 * @method static array|object concat(array|object $param, array|object $param) <p>Combines and concat collections provided with each others.</p><br /><p>If the collections have common keys, then the values are appended in an array.<br />If numerical indexes are passed, then values are appended.</p><br /><p>For a recursive merge, see __::merge.</p><br /><ul><br /><li>__::concat(['color' =&gt; ['favorite' =&gt; 'red', 5], 3], [10, 'color' =&gt; ['favorite' =&gt; 'green', 'blue']]);</li><br /><li>// &gt;&gt; ['color' =&gt; ['favorite' =&gt; ['green'], 5, 'blue'], 3, 10]</li><br /></ul>
 * @method static array|object concatDeep(array|object $param, array|object $param) <p>Recursively combines and concat collections provided with each others.</p><br /><p>If the collections have common keys, then the values are appended in an array.<br />If numerical indexes are passed, then values are appended.</p><br /><p>For a non-recursive concat, see __::concat.</p><br /><ul><br /><li>__::concatDeep(['color' =&gt; ['favorite' =&gt; 'red', 5], 3], [10, 'color' =&gt; ['favorite' =&gt; 'green', 'blue']]);</li><br /><li>// &gt;&gt; ['color' =&gt; ['favorite' =&gt; ['red', 'green'], 5, 'blue'], 3, 10]</li><br /></ul>
 * @method static null doForEach(array|object $param, \Closure $param) <p>Iterate over elements of the collection and invokes iteratee for each element.</p><br /><p>The iteratee is invoked with three arguments: (value, index|key, collection).<br />Iteratee functions may exit iteration early by explicitly returning false.</p><br /><ul><br /><li>__::doForEach([1, 2, 3], function ($value) { print_r($value) });</li><br /><li>// → (Side effect: print 1, 2, 3)</li><br /></ul>
 * @method static null doForEachRight(array|object $param, \Closure $param) <p>Iterate over elements of the collection, from right to left, and invokes iteratee<br />for each element.</p><br /><p>The iteratee is invoked with three arguments: (value, index|key, collection).<br />Iteratee functions may exit iteration early by explicitly returning false.</p><br /><ul><br /><li>__::doForEachRight([1, 2, 3], function ($value) { print_r($value) });</li><br /><li>// → (Side effect: print 3, 2, 1)</li><br /></ul>
 * @method static array ease(array $param, string $param) <p>Flattens a complex collection by mapping each ending leafs value to a key consisting of all previous indexes.</p><br /><p>__::ease(['foo' =&gt; ['bar' =&gt; 'ter'], 'baz' =&gt; ['b', 'z']]);<br />// → '['foo.bar' =&gt; 'ter', 'baz.0' =&gt; 'b', , 'baz.1' =&gt; 'z']'</p>
 * @method static void _ease(array $param, array $param, string $param, string $param) <p>Inner function for collections::ease</p><br />
 * @method static bool every(array|object $param, \Closure $param) <p>Checks if predicate returns truthy for all elements of collection.</p><br /><p>Iteration is stopped once predicate returns falsey.<br />The predicate is invoked with three arguments: (value, index|key, collection).</p><br /><ul><br /><li>__::every([1, 3, 4], function ($v) { return is_int($v); });</li><br /><li>// → true</li><br /></ul>
 * @method static array filter(array $param, \Closure|null $param) <p>Returns the values in the collection that pass the truth test.</p><br />
 * @method static array|mixed first(array $param, int|null $param) <p>Gets the first element of an array. Passing n returns the first n elements.</p><br />
 * @method static array|mixed|null get(array|object $param, string $param, \Closure|mixed $param) <p>get item of an array by index, accepting path (nested index).</p><br /><p>If $collection is an object that implementes the ArrayAccess interface, this<br />function will treat it as an array instead of accessing class properties.</p><br /><ul><br /><li>__::get(['foo' =&gt; ['bar' =&gt; 'ter']], 'foo.bar');</li><br /><li>// → 'ter'</li><br /></ul>
 * @method static array groupBy(array $param, int|float|string|\Closure $param) <p>Returns an associative array where the keys are values of $key.</p><br /><p>Based on {@author Chauncey McAskill}'s {@link <a href="https://gist.github.com/mcaskill/baaee44487653e1afc0d">https://gist.github.com/mcaskill/baaee44487653e1afc0d</a> array_group_by()}<br />function.</p><br /><ul><br /><li><br /><p>__::groupBy([</p><br /></li><br /><li><br /><p>['state' =&gt; 'IN', 'city' =&gt; 'Indianapolis', 'object' =&gt; 'School bus'],</p><br /></li><br /><li><br /><p>['state' =&gt; 'CA', 'city' =&gt; 'San Diego', 'object' =&gt; 'Light bulb'],</p><br /></li><br /><li><br /><p>['state' =&gt; 'CA', 'city' =&gt; 'Mountain View', 'object' =&gt; 'Space pen'],</p><br /></li><br /><li><br /><p>],</p><br /></li><br /><li><br /><p>'state'</p><br /></li><br /><li><br /><p>);</p><br /></li><br /><li><br /><p>// &gt;&gt; [</p><br /></li><br /><li><br /><p>//   'IN' =&gt; [</p><br /></li><br /><li><br /><p>//      ['state' =&gt; 'IN', 'city' =&gt; 'Indianapolis', 'object' =&gt; 'School bus'],</p><br /></li><br /><li><br /><p>//      ['state' =&gt; 'CA', 'city' =&gt; 'San Diego', 'object' =&gt; 'Light bulb'],</p><br /></li><br /><li><br /><p>//   ],</p><br /></li><br /><li><br /><p>//   'CA' =&gt; [</p><br /></li><br /><li><br /><p>//      ['state' =&gt; 'CA', 'city' =&gt; 'Mountain View', 'object' =&gt; 'Space pen']</p><br /></li><br /><li><br /><p>//   ]</p><br /></li><br /><li><br /><p>// ]</p><br /></li><br /><li><br /><p>__::groupBy([</p><br /></li><br /><li><br /><p>['object' =&gt; 'School bus', 'metadata' =&gt; ['state' =&gt; 'IN', 'city' =&gt; 'Indianapolis']],</p><br /></li><br /><li><br /><p>['object' =&gt; 'Manhole', 'metadata' =&gt; ['state' =&gt; 'IN', 'city' =&gt; 'Indianapolis']],</p><br /></li><br /><li><br /><p>['object' =&gt; 'Basketball', 'metadata' =&gt; ['state' =&gt; 'IN', 'city' =&gt; 'Plainfield']],</p><br /></li><br /><li><br /><p>['object' =&gt; 'Light bulb', 'metadata' =&gt; ['state' =&gt; 'CA', 'city' =&gt; 'San Diego']],</p><br /></li><br /><li><br /><p>['object' =&gt; 'Space pen', 'metadata' =&gt; ['state' =&gt; 'CA', 'city' =&gt; 'Mountain View']],</p><br /></li><br /><li><br /><p>],</p><br /></li><br /><li><br /><p>'metadata.state'</p><br /></li><br /><li><br /><p>);</p><br /></li><br /><li><br /><p>// &gt;&gt; [</p><br /></li><br /><li><br /><p>//   'IN' =&gt; [</p><br /></li><br /><li><br /><p>//     'Indianapolis' =&gt; [</p><br /></li><br /><li><br /><p>//       ['object' =&gt; 'School bus', 'metadata' =&gt; ['state' =&gt; 'IN', 'city' =&gt; 'Indianapolis']],</p><br /></li><br /><li><br /><p>//       ['object' =&gt; 'Manhole', 'metadata' =&gt; ['state' =&gt; 'IN', 'city' =&gt; 'Indianapolis']],</p><br /></li><br /><li><br /><p>//     ],</p><br /></li><br /><li><br /><p>//     'Plainfield' =&gt; [</p><br /></li><br /><li><br /><p>//       ['object' =&gt; 'Basketball', 'metadata' =&gt; ['state' =&gt; 'IN', 'city' =&gt; 'Plainfield']],</p><br /></li><br /><li><br /><p>//     ],</p><br /></li><br /><li><br /><p>//   ],</p><br /></li><br /><li><br /><p>//   'CA' =&gt; [</p><br /></li><br /><li><br /><p>//     'San Diego' =&gt; [</p><br /></li><br /><li><br /><p>//       ['object' =&gt; 'Light bulb', 'metadata' =&gt; ['state' =&gt; 'CA', 'city' =&gt; 'San Diego']],</p><br /></li><br /><li><br /><p>//     ],</p><br /></li><br /><li><br /><p>//     'Mountain View' =&gt; [</p><br /></li><br /><li><br /><p>//       ['object' =&gt; 'Space pen', 'metadata' =&gt; ['state' =&gt; 'CA', 'city' =&gt; 'Mountain View']],</p><br /></li><br /><li><br /><p>//     ],</p><br /></li><br /><li><br /><p>//   ],</p><br /></li><br /><li><br /><p>// ]</p><br /></li><br /><li><br /><p>__::groupBy([</p><br /></li><br /><li><br /><p>['state' =&gt; 'IN', 'city' =&gt; 'Indianapolis', 'object' =&gt; 'School bus'],</p><br /></li><br /><li><br /><p>['state' =&gt; 'IN', 'city' =&gt; 'Indianapolis', 'object' =&gt; 'Manhole'],</p><br /></li><br /><li><br /><p>['state' =&gt; 'CA', 'city' =&gt; 'San Diego', 'object' =&gt; 'Light bulb'],</p><br /></li><br /><li><br /><p>],</p><br /></li><br /><li><br /><p>function ($value) {</p><br /></li><br /><li><br /><p>return $value-&gt;city;</p><br /></li><br /><li><br /><p>}</p><br /></li><br /><li><br /><p>);</p><br /></li><br /><li><br /><p>// &gt;&gt; [</p><br /></li><br /><li><br /><p>//   'Indianapolis' =&gt; [</p><br /></li><br /><li><br /><p>//      ['state' =&gt; 'IN', 'city' =&gt; 'Indianapolis', 'object' =&gt; 'School bus'],</p><br /></li><br /><li><br /><p>//      ['state' =&gt; 'IN', 'city' =&gt; 'Indianapolis', 'object' =&gt; 'Manhole'],</p><br /></li><br /><li><br /><p>//   ],</p><br /></li><br /><li><br /><p>//   'San Diego' =&gt; [</p><br /></li><br /><li><br /><p>//      ['state' =&gt; 'CA', 'city' =&gt; 'San Diego', 'object' =&gt; 'Light bulb'],</p><br /></li><br /><li><br /><p>//   ]</p><br /></li><br /><li><br /><p>// ]</p><br /></li><br /></ul>
 * @method static bool has(array|object $param, string|int $param) <p>Return true if $collection contains the requested $key.</p><br /><p>In contrast to isset(), __::has() returns true if the key exists but is null.</p><br /><ul><br /><li><br /><p>__::has(['foo' =&gt; ['bar' =&gt; 'num'], 'foz' =&gt; 'baz'], 'foo.bar');</p><br /></li><br /><li><br /><p>// → true</p><br /></li><br /><li><br /><p>__::hasKeys((object) ['foo' =&gt; 'bar', 'foz' =&gt; 'baz'], 'bar');</p><br /></li><br /><li><br /><p>// → false</p><br /></li><br /></ul>
 * @method static bool hasKeys(array|object $param, array $param, bool $param) <p>Returns if $input contains all requested $keys. If $strict is true it also checks if $input exclusively contains the given $keys.</p><br /><ul><br /><li>__::hasKeys(['foo' =&gt; 'bar', 'foz' =&gt; 'baz'], ['foo', 'foz']);</li><br /><li>// → true</li><br /></ul>
 * @method static bool isEmpty(void $param) <p>Check if value is an empty array or object.</p><br /><p>We consider any non enumerable as empty.</p><br /><ul><br /><li>__::isEmpty([]);</li><br /><li>// → true</li><br /></ul>
 * @method static array|mixed last(array $param, null $param) <p>get last item(s) of an array</p><br />
 * @method static array map(array|object $param, \Closure $param) <p>Returns an array of values by mapping each in collection through the iteratee.</p><br /><p>The iteratee is invoked with three arguments: (value, index|key, collection).</p>
 * @method static array mapKeys(array $param, \Closure $param) <p>Transforms the keys in a collection by running each key through the iterator</p><br />
 * @method static array mapValues(array $param, \Closure $param) <p>Transforms the values in a collection by running each value through the iterator</p><br />
 * @method static array|object merge(array|object $param, array|object $param) <p>Recursively combines and merge collections provided with each others.</p><br /><p>If the collections have common keys, then the last passed keys override the previous.<br />If numerical indexes are passed, then last passed indexes override the previous.</p><br /><p>For a non-recursive merge, see __::merge.</p><br /><ul><br /><li>__::merge(['color' =&gt; ['favorite' =&gt; 'red', 'model' =&gt; 3, 5], 3], [10, 'color' =&gt; ['favorite' =&gt; 'green', 'blue']]);</li><br /><li>// &gt;&gt; ['color' =&gt; ['favorite' =&gt; 'green', 'model' =&gt; 3, 'blue'], 10]</li><br /></ul>
 * @method static array pick(array|object $param, array $param) <p>Returns an array having only keys present in the given path list.</p><br /><p>Values for missing keys values will be filled with provided default value.</p><br /><ul><br /><li>__::pick(['a' =&gt; 1, 'b' =&gt; ['c' =&gt; 3, 'd' =&gt; 4]], ['a', 'b.d']);</li><br /><li>// → ['a' =&gt; 1,'b' =&gt; ['d' =&gt; 4]]</li><br /></ul>
 * @method static array pluck(array|object $param, string $param) <p>Returns an array of values belonging to a given property of each item in a collection.</p><br />
 * @method static array|object set(array|object $param, string $param, mixed $param) <p>Return a new collection with the item set at index to given value.</p><br /><p>Index can be a path of nested indexes.</p><br /><p>If $collection is an object that implements the ArrayAccess interface, this<br />function will treat it as an array.</p><br /><p>If a portion of path doesn't exist, it's created. Arrays are created for missing<br />index in an array; objects are created for missing property in an object.</p><br /><ul><br /><li>__::set(['foo' =&gt; ['bar' =&gt; 'ter']], 'foo.baz.ber', 'fer');</li><br /><li>// → '['foo' =&gt; ['bar' =&gt; 'ter', 'baz' =&gt; ['ber' =&gt; 'fer']]]'</li><br /></ul>
 * @method static array unease(array $param, string $param) <p>Builds a multidimensional collection out of a hash map using the key as indicator where to put the value.</p><br /><ul><br /><li>__::unease(['foo.bar' =&gt; 'ter', 'baz.0' =&gt; 'b', , 'baz.1' =&gt; 'z']);</li><br /><li>// → '['foo' =&gt; ['bar' =&gt; 'ter'], 'baz' =&gt; ['b', 'z']]'</li>br /></ul>
 * @method static array where(array $param, array $param) <p>return data matching specific key value condition</p><br /><p>*_::where($a, ['age' =&gt; 16]);</p><br /><ul><br /><li>// &gt;&gt; [['name' =&gt; 'maciej', 'age' =&gt; 16]]</li><br /></ul>
 * @method static string slug(string $param, array $param) <p>Create a web friendly URL slug from a string.</p><br /><p>Although supported, transliteration is discouraged because<br />1) most web browsers support UTF-8 characters in URLs<br />2) transliteration causes a loss of information</p>
 * @method static string truncate(string $param, int $param) <p>trunct string based on count of words</p><br />
 * @method static string urlify(string $param) <ul><br /><li>__::urlify(&quot;I love <a href="https://google.com">https://google.com</a>&quot;);</li><br /><li>// → 'I love <a href="https://google.com">google.com</a>'</li><br /></ul><br />
 * @method static bool isCollection(null $param) <p>Check if the object is a collection.</p><br /><p>A collection is either an array or an object.</p>
 * @method static bool isEmail(null $param) <p>check if the value is valid email</p><br />
 * @method static mixed chain(mixed $param) <p>returns a wrapper instance, allows the value to be passed through multiple bottomline functions</p><br /><p>__::chain([0, 1, 2, 3, null])<br />-&gt;compact()<br />-&gt;prepend(4)<br />-&gt;value();</p><br /><blockquote><br /><blockquote><br /><p>[4, 1, 2, 3]</p><br /></blockquote><br /></blockquote>
 * @method static string camelCase(string $param) <p>Converts string to <a href="https://en.wikipedia.org/wiki/CamelCase">camel case</a>.</p><br /><p>__::camelCase('Foo Bar');</p><br /><blockquote><br /><blockquote><br /><p>'fooBar'</p><br /></blockquote><br /></blockquote>
 * @method static string capitalize(string $param) <p>Converts the first character of string to upper case and the remaining to lower case.</p><br /><p>__::capitalize('FRED');</p><br /><blockquote><br /><blockquote><br /><p>'Fred'</p><br /></blockquote><br /></blockquote>
 * @method static string kebabCase(string $param) <p>Converts string to<br /><a href="https://en.wikipedia.org/wiki/Letter_case#Special_case_styles">kebab case</a>.</p><br /><p>__::kebabCase('Foo Bar');</p><br /><blockquote><br /><blockquote><br /><p>'foo-bar'</p><br /></blockquote><br /></blockquote>
 * @method static string lowerCase(string $param) <p>Converts string, as space separated words, to lower case.</p><br /><p>__::lowerCase('--Foo-Bar--');</p><br /><blockquote><br /><blockquote><br /><p>'foo bar'</p><br /></blockquote><br /></blockquote>
 * @method static string lowerFirst(string $param) <p>Converts the first character of string to lower case, like lcfirst.</p><br /><p>__::lowerFirst('Fred');</p><br /><blockquote><br /><blockquote><br /><p>'fred'</p><br /></blockquote><br /></blockquote>
 * @method static string snakeCase(string $param) <p>Converts string to<br /><a href="https://en.wikipedia.org/wiki/Snake_case">snake case</a>.</p><br /><p>__::snakeCase('Foo Bar');</p><br /><blockquote><br /><blockquote><br /><p>'foo_bar'</p><br /></blockquote><br /></blockquote>
 * @method static string split(string $param, string $param, int $param) <p>Split a string by string.</p><br /><p>Based on explode, see <a href="http://php.net/manual/en/function.explode.php">http://php.net/manual/en/function.explode.php</a>.</p><br /><p>__::split('a-b-c', '-', 2);</p><br /><blockquote><br /><blockquote><br /><p>['a', 'b-c']</p><br /></blockquote><br /></blockquote>
 * @method static string startCase(string $param) <p>Converts string to <a href="https://en.wikipedia.org/wiki/Letter_case#Stylistic_or_specialised_usage">start case</a>.</p><br /><p>__::startCase('--foo-bar--');</p><br /><blockquote><br /><blockquote><br /><p>'Foo Bar'</p><br /></blockquote><br /></blockquote>
 * @method static string toLower(string $param) <p>Converts string, as a whole, to lower case just like strtolower.</p><br /><p>__::toLower('fooBar');</p><br /><blockquote><br /><blockquote><br /><p>'foobar'</p><br /></blockquote><br /></blockquote>
 * @method static string toUpper(string $param) <p>Converts string, as a whole, to lower case just like strtoupper.</p><br /><p>__::toUpper('fooBar');</p><br /><blockquote><br /><blockquote><br /><p>'FOOBAR'</p><br /></blockquote><br /></blockquote>
 * @method static string upperCase(string $param) <p>Converts string, as space separated words, to upper case.</p><br /><p>__::upperCase('--foo-bar');</p><br /><blockquote><br /><blockquote><br /><p>'FOO BAR'</p><br /></blockquote><br /></blockquote>
 * @method static string upperFirst(string $param) <p>Converts the first character of string to upper case, like ucfirst.</p><br /><p>__::upperFirst('fred');</p><br /><blockquote><br /><blockquote><br /><p>'Fred'</p><br /></blockquote><br /></blockquote>
 * @method static string words(string $param, string $param) <p>Splits string into an array of its words.</p><br /><p>__::words('fred, barney, &amp; pebbles');</p><br /><blockquote><br /><blockquote><br /><p>['fred', 'barney', 'pebbles']</p><br /></blockquote><br /></blockquote><br /><p>__::words('fred, barney, &amp; pebbles', '/[^, ]+/');</p><br /><blockquote><br /><blockquote><br /><p>['fred', 'barney', '&amp;', 'pebbles']</p><br /></blockquote><br /></blockquote>
 * @method static mixed identity() <p>Returns the first argument it receives</p><br /><p>__::identity('arg1', 'arg2');</p><br /><blockquote><br /><blockquote><br /><p>'arg1'</p><br /></blockquote><br /></blockquote>
 */
```

This would require a **thorough** clean up of the phpDocs for each function, but it'll allow for the documentation to be maintained in a single location consistently.

The `phpDocGen.php` script requires PHP 5.5.9 or higher (if I'm not mistaken) to be run due to dependency constraints. But since maintainers would be executing this script whenever new functions are added/updated, it won't bump the min version of 5.4.0 for actual library use.

This is just a large change to how documentation would be maintained, so it's not breaking backwards compatibility.

I'd love to hear your thoughts @MonsieurV and @maciejczyzewski

### CODE REVIEW

**Comment by MonsieurV** [comment]
100% for using the script to generate the phpDoc: all that is currently in the `load.php` and is indeed not very DRY to maintain.

For the README, seems a step further. I'm not sure we want exactly the same thing in our web/README documentation than in our phpDoc and in-code documentation. Many times the README doc is shorter, getting to the point more rapidly, while the phpDoc show the use cases more thoroughly.

Anyway if it can already reduce duplication from 3 to 2 it is already a great step forward. :+1:

**Comment by MonsieurV** [comment]
For the PHP 5.4 test that fails: seems PHPUnit try to compute coverage on the `reflection-docblock` package.

https://travis-ci.org/maciejczyzewski/bottomline/jobs/382027705#L547

Can't it be black listed or something? (In the stack trace we see PHPUnit calling black list filters just before trying to load the class)

**Comment by allejo** [comment]
> For the README, seems a step further. I'm not sure we want exactly the same thing in our web/README documentation than in our phpDoc and in-code documentation. Many times the README doc is shorter, getting to the point more rapidly, while the phpDoc show the use cases more thoroughly.

Doc blocks allow for a "summary" and a "description". Maybe as the doc blocks are rewritten, the summaries remain short and concise. Then it's *just* the summaries that are included in the README along with any code samples extracted?

> Can't it be black listed or something? (In the stack trace we see PHPUnit calling black list filters just before trying to load the class)

Tried this a number of different ways and I wasn't successful. I believe it's Composer's autoloading that is picking up those classes before phpunit. Anyhow, I updated the version requirements for `reflection-docblock` to pick either `2.x` or `3.x` based on the current PHP version. Since the script runs with a min of `5.5.x`, there shouldn't be any issues.

**Comment by allejo** [comment]
Before I proceed to tackle the rest of the docs for the other namespaces, here's a sample of what I did to the `arrays` namespace and how it's generated for `load.php`.

```php
/**
 * 
 *
 * @method static array append(array $array, mixed $value) <p>Append an item to array.</p><br><p><strong>Usage</strong></p><pre><code class="language-php">__::append([1, 2, 3], 4);</code></pre><p><strong>Result</strong></p><pre><code>[1, 2, 3, 4]</code></pre>
 * @method static array chunk(array $array, int $size, bool $preserveKeys) <p>Creates an array of elements split into groups the length of <code>$size</code>.</p><br><p>If array can't be split evenly, the final chunk will be the remaining elements.When <code>$preserveKeys</code> is set to TRUE, keys will be preserved. Default is FALSE,which will reindex the chunk numerically.</p><p><strong>Usage</strong></p><pre><code class="language-php">__::chunk([1, 2, 3, 4, 5], 3);</code></pre><p><strong>Result</strong></p><pre><code>[[1, 2, 3], [4, 5]]</code></pre>
 * @method static array compact(array $array) <p>Creates an array with all falsey values removed.</p><br><p>The following values are considered falsey:</p><ul><li><code>false</code></li><li><code>null</code></li><li><code>0</code></li><li><code>""</code></li><li><code>undefined</code></li><li><code>NaN</code></li></ul><p><strong>Usage</strong></p><pre><code class="language-php">__::compact([0, 1, false, 2, '', 3]);</code></pre><p><strong>Result</strong></p><pre><code>[1, 2, 3]</code></pre>
 * @method static array drop(array $input, int $number) <p>Creates a slice of array with n elements dropped from the beginning.</p><br><p><strong>Usage</strong></p><pre><code class="language-php">__::drop([0, 1, 3, 5], 2);</code></pre><p><strong>Result</strong></p><pre><code>[3, 5]</code></pre>
 * @method static array flatten(array $array, bool $shallow) <p>Flattens a multidimensional array.</p><br><p>If <code>$shallow</code> is set to TRUE, the array will only be flattened a single level.</p><p><strong>Usage</strong></p><pre><code class="language-php">__::flatten([1, 2, [3, [4]]], false);</code></pre><p><strong>Result</strong></p><pre><code>[1, 2, 3, 4]</code></pre>
 * @method static array patch(array $array, array $patches, string $parent) <p>Patches array by xpath.</p><br><p><strong>Usage</strong></p><pre><code class="language-php">__::patch(<br />    [<br />        'addr' =&gt; [<br />            'country' =&gt; 'US',<br />            'zip' =&gt; 12345<br />        ]<br />    ],<br />    [<br />        '/addr/country' =&gt; 'CA',<br />        '/addr/zip' =&gt; 54321<br />    ]<br />);</code></pre><p><strong>Result</strong></p><pre><code>['addr' =&gt; ['country' =&gt; 'CA', 'zip' =&gt; 54321]]</code></pre>
 * @method static array prepend(array $array, mixed $value) <p>Prepend item or value to an array.</p><br><p><strong>Usage</strong></p><pre><code class="language-php">__::prepend([1, 2, 3], 4);</code></pre><p><strong>Result</strong></p><pre><code>[4, 1, 2, 3]</code></pre>
 * @method static array randomize(array $array) <p>Shuffle an array ensuring no item remains in the same position.</p><br><p><strong>Usage</strong></p><pre><code class="language-php">__::randomize([1, 2, 3]);</code></pre><p><strong>Result</strong></p><pre><code>[2, 3, 1]</code></pre>
 * @method static array range(int $start, int $stop, int $step) <p>Generate range of values based on start, end, and step.</p><br><p><strong>Usage</strong></p><pre><code class="language-php">__::range(1, 10, 2);</code></pre><p><strong>Result</strong></p><pre><code>[1, 3, 5, 7, 9]</code></pre>
 * @method static array repeat(string $object, int $times) <p>Generate array of repeated values.</p><br><p><strong>Usage</strong></p><pre><code class="language-php">__::repeat('foo', 3);</code></pre><p><strong>Result</strong></p><pre><code>['foo', 'foo', 'foo']</code></pre>
 */
```

Thoughts or suggestions?

**Comment by MonsieurV** [comment]
Seems very good for me! :+1: 

> Then it's just the summaries that are included in the README along with any code samples extracted?

Yes, could be that. Maybe I was thinking we are not showing too much samples in the README to not make it too hard to navigate.
However at this point, based on your work, we could go further and generate an HTML page for the main documentation. This would alleviate presentation issues, as we could add a function index, a search and clearly differentiate code samples from the summaries (so we can have more without cluttering too much the doc navigation).

I'll be happy to contribute to that. In the meantime let's first have the phpDoc generated as the basis: you're doing great.

**Comment by allejo** [comment]
All of the docs have been rewritten ✅ 

So all that's left is to get `phpDocGen.php` to correctly handle PHP files that only return a function name.

**Comment by allejo** [comment]
Another small annoyance is that `phpDocGen.php` will delete any whitespace between method definitions; see b5dfc024cbba1c073d5233ae846c5b6cae0fe4ae. Not sure if you'd want to handle it differently.

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -1,42 +1,50 @@
...
-
-## Recommendation
-- Squash commits into 1 commit before the PR is merged into master, this help reduces git tree and makes it easier to revert to a certain state. (can be multiple commits, but they have to be meaningful commits)
-- Open PR title starting with one of the following words:
-    + `Fix` - for a bug fix.
-    + `New` - implemented a new feature.
-    + `Update` - for a backwards-compatible enhancement.
-    + `Breaking` - for a backwards-incompatible enhancement or feature.
-    + `Docs` - changes to documentation only.
+# Contributing
```
Comment (MonsieurV) [inline CONTRIBUTING.md]: Not sure of what is changed in this file.

**Comment by MonsieurV** [comment]
> Another small annoyance is that phpDocGen.php will delete any whitespace between method definitions

Not sure this is an issue, as the file is to be generated and only read in a phpDoc format.

If we want to handle that, we could extract the category/theme from the directory, pass it to `registerBottomlineFunction` and store it to `bottomlineMethods` for generating the output.

File: CONTRIBUTING.md (hunk 1 above)
Comment (allejo) [reply CONTRIBUTING.md]: pfft. I didn't realize that line endings for that file probably changed.

```
$ file CONTRIBUTING.md
CONTRIBUTING.md: ASCII text, with very long lines, with CRLF, LF line terminators
```

Don't know if you'd like to change that later on.

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -1,38 +1,46 @@
...
-## Development checklist
+## Development checklist
+
+- Add or update phpDocs for the new function with a **Usage** and **Result** section
 - Always add tests for the code that you write, including edge cases
 - Place the new functions where they belong (collections, arrays, utilities, etc.)
 - Add `README.md` doc entry for the functions
-- Add function signatures to `__/load.php`
-- Update the benchmark `bench.php`, this helps us to validate the performance of the library
+- Execute the `phpDocGen.php` script to automatically build an updated `load.php`
```
Comment (MonsieurV) [inline CONTRIBUTING.md]: Will see if this could be unversioned and be the responsibility of the one pushing the build to Composer.

But KISS - and also it may be better to have it anyway directly in the GitHub repository (for the one using the GitHub repo directly in their `composer.json`)

**Comment by MonsieurV** [comment]
Lot of work here!

I permit myself to merge, and let @maciejczyzewski publish - so people can tell us how the phpDoc feels. :)

File: CONTRIBUTING.md (hunk 1 above)
Comment (MonsieurV) [reply CONTRIBUTING.md]: Oh just way it. Well just `LF`s is indeed better. (Or whatever we use in other files, but consistently)

I'll try to look at that... and on the same occasion run the [PHP-CS Fixer](https://github.com/FriendsOfPHP/PHP-CS-Fixer) on the project.

**Comment by allejo** [comment]
🎉 🎉 🎉

**Comment by MonsieurV** [comment]
@maciejczyzewski could we release?

We have much in the _unreleased_
https://github.com/maciejczyzewski/bottomline/blob/master/CHANGELOG.md#unreleased

(I was trying to use `assign`)

**Comment by allejo** [comment]
hey @maciejczyzewski, hope you don't mind a gentle nudge for a new tag ❤️
